File: apps/ios/ExpCore/Sources/Domain/PrGraph.swift
import Foundation

/// EXP-897 Part 4 — ONE model for everything a piece of work is entangled
/// with, mirrored ×4 (web `lib/pr-graph.ts`, desktop `pr_graph.rs`, Android
/// `PrGraph.kt`). The Work header's badge and its overlay read nothing else.
///
/// Three relationships, all derived from ALREADY SYNCED rows — no new column,
/// no request:
///
/// - **stack**: pull requests chained by `prBaseBranch → branch` (`PrStack`).
/// - **batch**: the issues sharing ONE `prUrl` — a batch run's combined PR.
///   A batch is an ENTRY like any other, so a batch can be a stack member.
/// - **tree**: the session tree off `parentSessionId` (`SessionTree`).
///
/// Plus the issue's OPEN blockers (`StackStart.openBlockers`), which is what
/// the Issue face's "Blocked by" section lists.
public enum PrGraph {

    /// ONE pull request: a single issue, or the issues of a batch PR.
    public struct Entry: Identifiable, Equatable {
        /// `prUrl` when there is one, else `issue:<id>` — the grouping key.
        public let id: String
        /// The issues on this PR, newest first; `representative` is the first.
        public let issues: [IssueEntity]

        public init(id: String, issues: [IssueEntity]) {
            self.id = id
            self.issues = issues
        }

        public var representative: IssueEntity { issues[0] }
        public var isBatch: Bool { issues.count > 1 }
        public var prUrl: String? { representative.prUrl }
        public var prNumber: Int? { representative.prNumber }
        public var prState: String? { representative.prState }
        public var branch: String? { representative.branch }
        public var prBaseBranch: String? { representative.prBaseBranch }
        public var identifiers: [String] { issues.compactMap(\.identifier) }

        public static func == (a: Entry, b: Entry) -> Bool { a.id == b.id }
    }

    /// One rung of the stack, bottom first.
    public struct StackEntry: Identifiable {
        public let entry: Entry
        /// 0 at the BOTTOM of the chain, +1 per rung up.
        public let depth: Int
        public var id: String { entry.id }
    }

    /// The batch the subject is part of — the issues sharing its PR.
    public struct BatchEntry {
        public let issues: [IssueEntity]
    }

    /// What the header badge says.
    public enum BadgeKind: Equatable {
        case stack
        case batch
        case stackAndBatch
    }

    public struct Graph {
        /// The chain this work sits in, bottom first. Empty when it is not
        /// stacked (a lone PR is not a stack).
        public let stack: [StackEntry]
        /// The batch this work rides, or nil when its PR is its own.
        public let batch: BatchEntry?
        /// The run tree the subject's session belongs to (roots first).
        public let tree: [SessionTree.Row<CodingSessionEntity>]
        /// The issue's OPEN blockers — the Issue face's "Blocked by" list.
        public let blockers: [IssueEntity]
        /// The subject's own entry, when it has one.
        public let entry: Entry?

        /// The subject's 1-based rung, counted from the bottom. Nil unstacked.
        public var position: Int? {
            guard let entry, !stack.isEmpty else { return nil }
            guard let index = stack.firstIndex(where: { $0.entry.id == entry.id }) else {
                return nil
            }
            return index + 1
        }

        public var size: Int { stack.count }

        /// `2 of 3` — the badge's caption, nil when there is no stack.
        public var positionLabel: String? {
            guard let position, size > 1 else { return nil }
            return "\(position) of \(size)"
        }

        /// The entry directly below the subject (its foundation).
        public var below: Entry? {
            guard let position, position > 1 else { return nil }
            return stack[position - 2].entry
        }

        /// The entry directly above the subject.
        public var above: Entry? {
            guard let position, position < size else { return nil }
            return stack[position].entry
        }

        /// The BOTTOM of the chain — where "Merge stack" lives.
        public var bottom: Entry? { stack.first?.entry }

        /// Nothing to show: no stack, no batch, no tree, no blockers.
        public var isEmpty: Bool {
            stack.count < 2 && batch == nil && tree.count < 2 && blockers.isEmpty
        }
    }

    /// Build the graph for a subject: an issue, a session, or both (the Work
    /// screen usually has both). `issues` is every synced issue in scope,
    /// `sessions` every synced session, `relations` every synced relation row.
    public static func build(
        issue: IssueEntity?,
        session: CodingSessionEntity?,
        issues: [IssueEntity],
        sessions: [CodingSessionEntity],
        relations: [IssueRelationEntity]
    ) -> Graph {
        let subject = issue ?? session?.issueId.flatMap { id in issues.first { $0.id == id } }
        let entries = prEntries(issues)
        let subjectEntry = subject.flatMap { subject in
            entries.first { entry in entry.issues.contains { $0.id == subject.id } }
        } ?? batchSessionEntry(session, issues: issues, entries: entries)

        let chain: [StackEntry]
        if let subjectEntry {
            let members = PrStack.stackChain(
                subjectEntry,
                in: entries,
                id: { $0.id },
                branch: { $0.branch },
                base: { $0.prBaseBranch }
            )
            chain = members.count > 1
                ? members.enumerated().map { StackEntry(entry: $0.element, depth: $0.offset) }
                : []
        } else {
            chain = []
        }

        let batch = subjectEntry.flatMap { $0.isBatch ? BatchEntry(issues: $0.issues) : nil }

        let blockers = subject.map {
            StackStart.openBlockers(issueId: $0.id, relations: relations, issues: issues)
        } ?? []

        return Graph(
            stack: chain,
            batch: batch,
            tree: sessionTree(session: session, sessions: sessions),
            blockers: blockers,
            entry: subjectEntry
        )
    }

    /// The badge a graph earns: a stack, a batch, both, or nothing at all.
    public static func badgeKind(_ graph: Graph) -> BadgeKind? {
        let stacked = graph.stack.count > 1
        let batched = graph.batch != nil
        if stacked && batched { return .stackAndBatch }
        if stacked { return .stack }
        if batched { return .batch }
        return nil
    }

    /// EXP-1058: what the header's STACKED issue chip draws in place of the
    /// old pill — the front chip's issue and how many ride behind it (`+N`).
    public struct BadgeChip: Equatable {
        /// The subject PR's representative row; nil only for a run family
        /// with no issue (the front chip names the run instead).
        public let issue: IssueEntity?
        /// Every OTHER issue on the stack (all entries) or batch, or every
        /// other run of the tree for a runs-only family.
        public let count: Int

        public static func == (a: BadgeChip, b: BadgeChip) -> Bool {
            a.issue?.id == b.issue?.id && a.count == b.count
        }
    }

    /// The stacked chip for `face`, nil exactly when there is no badge: a
    /// stack/batch on every face, else a run tree (2+ runs) on the Run face
    /// (Results = the run's face). Mirrors web `badgeChip` (`pr-graph.ts`),
    /// desktop `pr_graph::badge_chip`, Android `PrGraph.badgeChip`.
    public static func badgeChip(_ graph: Graph, face: WorkFaceKind) -> BadgeChip? {
        let issue = graph.entry?.representative
        guard badgeKind(graph) != nil else {
            let runFace = face == .run || face == .results
            guard runFace, graph.tree.count > 1 else { return nil }
            return BadgeChip(issue: issue, count: graph.tree.count - 1)
        }
        if graph.stack.count >= 2 {
            let total = graph.stack.reduce(0) { $0 + $1.entry.issues.count }
            return BadgeChip(issue: issue, count: total - 1)
        }
        return BadgeChip(issue: issue, count: (graph.batch?.issues.count ?? 1) - 1)
    }

    // MARK: - Pieces

    /// Every pull request among `issues`, one entry per distinct `prUrl`
    /// (issues sharing one = a batch), an issue with no PR keyed on its own
    /// id. First-seen order, newest issue first inside an entry — the Reviews
    /// rule, so the overlay and the list read the same.
    /// EXP-876: a BATCH run's own entry. A batch links no issue and stamps no
    /// `pr_url` of its own, so before this it resolved nothing at all — the
    /// pill and its sheet, the one surface built to name work that spans
    /// several issues, never appeared on the very run that spans them. Its
    /// covered set (`batch_issue_ids`) IS the entry.
    ///
    /// The PR-grouped entry wins whenever there is one: it carries the branch
    /// and the base the stack chains on, so a batch PR stacked on another
    /// still reads `stack+batch` and still offers Merge stack. The synthesized
    /// entry is what a batch wears BEFORE its PR exists.
    static func batchSessionEntry(
        _ session: CodingSessionEntity?,
        issues: [IssueEntity],
        entries: [Entry]
    ) -> Entry? {
        guard let session else { return nil }
        let covered = BatchRun.issues(session, issues: issues)
        guard let first = covered.first else { return nil }
        if let grouped = entries.first(where: { entry in
            entry.isBatch && entry.issues.contains { $0.id == first.id }
        }) {
            return grouped
        }
        return Entry(id: "run:\(session.id)", issues: covered)
    }

    public static func prEntries(_ issues: [IssueEntity]) -> [Entry] {
        var buckets: [String: [IssueEntity]] = [:]
        var keyOrder: [String] = []
        for issue in issues {
            let key = (issue.prUrl?.isEmpty == false) ? issue.prUrl! : "issue:\(issue.id)"
            if buckets[key] == nil {
                keyOrder.append(key)
                buckets[key] = []
            }
            buckets[key]?.append(issue)
        }
        return keyOrder.compactMap { key in
            guard let bucket = buckets[key], !bucket.isEmpty else { return nil }
            return Entry(id: key, issues: bucket.sorted(by: newerFirst))
        }
    }

    /// The run tree the subject's session belongs to: its whole ROOT subtree,
    /// nested. Empty without a session.
    private static func sessionTree(
        session: CodingSessionEntity?, sessions: [CodingSessionEntity]
    ) -> [SessionTree.Row<CodingSessionEntity>] {
        guard let session else { return [] }
        let byId = Dictionary(sessions.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        // Up to the root (defensively bounded by the row count).
        var root = byId[session.id] ?? session
        var seen: Set<String> = [root.id]
        while let parentId = root.parentSessionId, let parent = byId[parentId],
              !seen.contains(parentId) {
            seen.insert(parentId)
            root = parent
        }
        let rows = SessionTree.nest(sessions)
        guard let start = rows.firstIndex(where: { $0.session.id == root.id }) else {
            return SessionTree.nest([root])
        }
        var subtree = [rows[start]]
        for row in rows[(start + 1)...] {
            if row.depth <= rows[start].depth { break }
            subtree.append(row)
        }
        return subtree
    }

    /// Newest-first by `createdAt` (Postgres wire text compares
    /// chronologically), id as the deterministic tie-break.
    private static func newerFirst(_ a: IssueEntity, _ b: IssueEntity) -> Bool {
        if a.createdAt != b.createdAt { return a.createdAt > b.createdAt }
        return a.id > b.id
    }
}

File: apps/web/src/lib/pr-graph.ts
// EXP-897 Part 4: the ONE model behind the stack/batch BADGE and its overlay.
// A piece of work can be related to other work three ways, and every client
// draws the same pill and the same three sections off this one function
// (iOS `PrGraph.swift`, Android `PrGraph.kt`, desktop `pr_graph.rs`):
//
//   · a STACK   — pull requests based on each other (`issues.pr_base_branch`)
//   · a BATCH   — issues sharing ONE pull request (`issues.pr_url`)
//   · a TREE    — runs that started other runs (`coding_sessions.parent_session_id`)
//
// A stack MEMBER is a pull request, not an issue, so a batch PR can itself sit
// in a stack — hence the entry type below. Nothing new is stored: every edge
// is already synced.
//
// Test names, mirrored ×4: `reports a stack badge for a stacked pr`,
// `reports a batch badge for a batch pr`, `reports both for a batch inside a
// stack`, `reports nothing for a lone pr`.

import { nestSessions, type SessionTreeRow, type TreeSession } from "@/lib/session-tree"
import { stackChain, type PrStackNode } from "@/lib/pr-stack"
import { openBlockers, type StackStartRelation } from "@/lib/stack-start"
import { batchRunIssues, isBatchRun, type BatchRunIssue } from "@/lib/batch-run"

export interface PrGraphIssue extends PrStackNode, BatchRunIssue {
  identifier: string
  /** The dual-written ANCHOR enum — what `openBlockers` judges. */
  status: string
  prUrl: string | null
}

export interface PrGraphSession extends TreeSession {
  issueId: string | null
  prUrl?: string | null
  /** EXP-876: a BATCH run's own subject — the issues it covers. Absent on a
   *  caller that does not carry them; such a run resolves no batch. */
  actionName?: string | null
  batchIssueIds?: string[] | null
  branch?: string | null
}

/** ONE pull request: a single issue, or every issue sharing its `prUrl`. */
export interface PrGraphEntry<I> {
  key: string
  /** The representative row — it carries the branch, base and PR fields. */
  issue: I
  /** Every issue on this pull request, in caller order (length 1 = plain). */
  issues: I[]
}

export interface PrGraph<I, S> {
  /** The pull request the subject belongs to; null when it has none. */
  entry: PrGraphEntry<I> | null
  /** The chain BOTTOM first, depth = distance from the bottom. Empty when the
   *  subject is in no stack. */
  stack: { entry: PrGraphEntry<I>; depth: number }[]
  /** The issues sharing the subject's pull request; null when it is not one. */
  batch: { issues: I[] } | null
  /** The subject run's whole tree (its root and every descendant), nested. */
  tree: SessionTreeRow<S>[]
  /** The issue face's "Blocked by" section — open blockers of the subject. */
  blockedBy: I[]
  /** 1-based position in `stack` (0 when not stacked) and the stack's size. */
  position: number
  size: number
}

export type BadgeKind = `stack` | `batch` | `stack+batch` | null

function groupEntries<I extends PrGraphIssue>(
  issues: readonly I[]
): { entries: PrGraphEntry<I>[]; byIssueId: Map<string, PrGraphEntry<I>> } {
  const entries: PrGraphEntry<I>[] = []
  const byKey = new Map<string, PrGraphEntry<I>>()
  const byIssueId = new Map<string, PrGraphEntry<I>>()
  for (const issue of issues) {
    // An issue with no pull request cannot collide — keyed by its own id.
    const key = issue.prUrl && issue.prUrl.length > 0 ? issue.prUrl : issue.id
    let entry = byKey.get(key)
    if (!entry) {
      entry = { key, issue, issues: [] }
      byKey.set(key, entry)
      entries.push(entry)
    }
    entry.issues.push(issue)
    byIssueId.set(issue.id, entry)
  }
  return { entries, byIssueId }
}

/**
 * EXP-876: a BATCH run's own entry. A batch links no issue and stamps no
 * `pr_url` of its own, so before this it resolved nothing at all — the pill
 * and its sheet, the one surface built to name work that spans several
 * issues, never appeared on the very run that spans them. Its covered set
 * (`batch_issue_ids`) IS the entry.
 *
 * The PR-grouped entry wins whenever there is one: it carries the branch and
 * the base the stack chains on, so a batch PR stacked on another still reads
 * `stack+batch` and still offers Merge stack. The synthesized entry is what a
 * batch wears BEFORE its PR exists — keyed by the run, since it has no url.
 */
function batchSessionEntry<I extends PrGraphIssue, S extends PrGraphSession>(
  session: S | null,
  issues: readonly I[],
  byIssueId: Map<string, PrGraphEntry<I>>
): PrGraphEntry<I> | null {
  if (!session || !isBatchRun({ issueId: session.issueId, actionName: session.actionName ?? null })) {
    return null
  }
  const covered = batchRunIssues(
    {
      issueId: session.issueId,
      actionName: session.actionName ?? null,
      batchIssueIds: session.batchIssueIds ?? null,
    },
    issues
  )
  const first = covered[0]
  if (!first) return null
  const grouped = byIssueId.get(first.id)
  if (grouped && grouped.issues.length > 1) return grouped
  return { key: `run:${session.id}`, issue: first, issues: covered }
}

/** The subtree ids of `rootId` plus the root itself, over raw sessions. */
function subtreeIds<S extends PrGraphSession>(
  sessions: readonly S[],
  rootId: string
): Set<string> {
  const childrenOf = new Map<string, S[]>()
  for (const session of sessions) {
    const parent = session.parentSessionId
    if (!parent || parent === session.id) continue
    const list = childrenOf.get(parent) ?? []
    list.push(session)
    childrenOf.set(parent, list)
  }
  const ids = new Set<string>([rootId])
  const queue = [rootId]
  while (queue.length > 0) {
    const id = queue.shift()!
    for (const child of childrenOf.get(id) ?? []) {
      if (ids.has(child.id)) continue
      ids.add(child.id)
      queue.push(child.id)
    }
  }
  return ids
}

/** The topmost ancestor of `session` present in `sessions` (itself when it has
 *  no listed parent). Cycle-safe. */
function rootOf<S extends PrGraphSession>(
  sessions: readonly S[],
  session: S
): S {
  const byId = new Map(sessions.map((row) => [row.id, row]))
  const seen = new Set<string>([session.id])
  let current = session
  for (;;) {
    const parentId = current.parentSessionId
    if (!parentId || seen.has(parentId)) return current
    const parent = byId.get(parentId)
    if (!parent) return current
    seen.add(parent.id)
    current = parent
  }
}

/**
 * Everything the badge and its overlay need for ONE subject — an issue, a run,
 * or both (a run's Issue face). Pure: every input is already synced.
 */
export function prGraph<
  I extends PrGraphIssue,
  S extends PrGraphSession,
>(input: {
  issue?: I | null
  session?: S | null
  issues: readonly I[]
  sessions: readonly S[]
  relations?: readonly StackStartRelation[]
}): PrGraph<I, S> {
  const { issues, sessions, relations = [] } = input
  const { entries, byIssueId } = groupEntries(issues)

  const subjectIssue =
    input.issue ??
    (input.session?.issueId
      ? (issues.find((row) => row.id === input.session!.issueId) ?? null)
      : null)

  const subjectEntry =
    (subjectIssue ? (byIssueId.get(subjectIssue.id) ?? null) : null) ??
    (input.session?.prUrl
      ? (entries.find((entry) => entry.key === input.session!.prUrl) ?? null)
      : null) ??
    batchSessionEntry(input.session ?? null, issues, byIssueId)

  // A stack walks over ENTRIES: the representatives carry the branch and the
  // base, and every issue of a batch PR shares them.
  const stack: { entry: PrGraphEntry<I>; depth: number }[] = []
  let position = 0
  let size = 0
  if (subjectEntry) {
    const representatives = entries.map((entry) => entry.issue)
    const chain = stackChain(subjectEntry.issue, representatives)
    if (chain.length >= 2) {
      size = chain.length
      chain.forEach((member, depth) => {
        const entry = byIssueId.get(member.id)
        if (!entry) return
        stack.push({ entry, depth })
        if (entry.key === subjectEntry.key) position = depth + 1
      })
    }
  }

  const batch =
    subjectEntry && subjectEntry.issues.length > 1
      ? { issues: subjectEntry.issues }
      : null

  let tree: SessionTreeRow<S>[] = []
  if (input.session) {
    const root = rootOf(sessions, input.session)
    const ids = subtreeIds(sessions, root.id)
    tree = nestSessions(sessions.filter((row) => ids.has(row.id)))
  } else if (subjectIssue) {
    const ids = new Set<string>()
    for (const row of sessions) {
      if (row.issueId !== subjectIssue.id) continue
      for (const id of subtreeIds(sessions, row.id)) ids.add(id)
    }
    tree = nestSessions(sessions.filter((row) => ids.has(row.id)))
  }

  const blockedBy = subjectIssue
    ? openBlockers(subjectIssue.id, relations, issues)
    : []

  return { entry: subjectEntry, stack, batch, tree, blockedBy, position, size }
}

/** Which glyph(s) the pill wears — `null` = no pill at all. */
export function badgeKind<I, S>(graph: PrGraph<I, S>): BadgeKind {
  const stacked = graph.stack.length >= 2
  const batched = graph.batch !== null
  if (stacked && batched) return `stack+batch`
  if (stacked) return `stack`
  if (batched) return `batch`
  return null
}

/** Which face the badge is drawn on — it decides the overlay's sections and,
 *  on the Run face, whether the session tree alone earns a pill. */
export type PrGraphFace = `issue` | `run` | `changes`

/** EXP-1079: what the header pill DRAWS — a PR relation (`badgeKind`), or,
 *  on the Run face of a run that has a family, the session tree alone
 *  (`runs`: the desktop's `BadgeGlyph::Runs`, the `session-tree` concept).
 *  `null` = no pill. The desktop twin is `BadgeSpec::is_visible` +
 *  `badge_glyphs`: there a lone run has an EMPTY tree, here the tree carries
 *  the subject itself, so "a family" is more than one row. */
export type BadgeShape = Exclude<BadgeKind, null> | `runs` | null

export function badgeShape<I, S>(
  graph: PrGraph<I, S>,
  face: PrGraphFace
): BadgeShape {
  const kind = badgeKind(graph)
  if (kind) return kind
  return face === `run` && graph.tree.length > 1 ? `runs` : null
}

/** The pill's own label: the stack position when there is one, else the
 *  batch's size. Byte-identical ×4. */
export function badgeLabel<I, S>(graph: PrGraph<I, S>): string | null {
  if (graph.stack.length >= 2) return `${graph.position} of ${graph.size}`
  if (graph.batch) {
    const count = graph.batch.issues.length
    return count === 1 ? `1 issue` : `${count} issues`
  }
  return null
}
